Unix/http/httpclient.c

```diff
@@ -1843,6 +1899,266 @@ static Page* _CreateHttpHeader(
     return page;
 }
 
+
+/**
+ * _UnpackDestinationOptions 
+ *
+ * returns: 
+ *   pUserName and pPassword are from a credential. That is only deleted at the end of the session
+ *   pTrustedCertsDir, pCertsFile, pPrivateKeyfile return allocated data that must be freed
+ *
+ **/
+MI_Result _UnpackDestinationOptions(
+    _In_ MI_DestinationOptions *pDestOptions,
+    _Out_opt_ AuthMethod *pAuthType,
+    _Out_opt_ char **pUsername,
+    _Out_opt_ char **pPassword,
+    _Out_opt_ MI_Uint32 *pPasswordLen,
+    _Out_opt_ char **pTrustedCertsDir,
+    _Out_opt_ char **pCertFile,
+    _Out_opt_ char **pPrivateKeyFile )
+
+{
+  static const MI_Char   AUTH_NAME_BASIC[]   = MI_AUTH_TYPE_BASIC;
+  //static const MI_Uint32 AUTH_NAME_BASIC_LEN = sizeof(AUTH_NAME_BASIC)/sizeof(MI_Char);
+
+  static const MI_Char   AUTH_NAME_NEGOTIATE[] = MI_AUTH_TYPE_NEGO_NO_CREDS;
+  //static const MI_Uint32 AUTH_NAME_NEGOTIATE_LEN = sizeof(AUTH_NAME_NEGOTIATE)/sizeof(MI_Char);
+
+  static const MI_Char   AUTH_NAME_NEGOTIATE_WITH_CREDS[] = MI_AUTH_TYPE_NEGO_WITH_CREDS;
+  //static const MI_Uint32 AUTH_NAME_NEGOTIATE_WITH_CREDS_LEN = sizeof(AUTH_NAME_NEGOTIATE_WITH_CREDS)/sizeof(MI_Char);
+
+  static const MI_Char   AUTH_NAME_KERBEROS[]   = MI_AUTH_TYPE_KERBEROS;
+  //static const MI_Uint32 AUTH_NAME_KERBEROS_LEN = sizeof(AUTH_NAME_KERBEROS)/sizeof(MI_Char);
+
+  MI_Uint32 cred_count = 0;
+  MI_Char *optionName = NULL;
+  MI_UserCredentials userCredentials;
+  MI_Uint32 username_len = 0;
+  MI_Uint32 password_len = 0;
+  char *username = NULL;
+  char *password = NULL;
+  MI_Result result = MI_RESULT_OK;
+
+  AuthMethod method = AUTH_METHOD_NONE;
+
+   // Unpack the destination options
+   /* Must have one and only one credential */
+
+    if ((MI_DestinationOptions_GetCredentialsCount(pDestOptions, &cred_count) != MI_RESULT_OK) ||
+            (cred_count != 1))
+    {
+        result = MI_RESULT_ACCESS_DENIED;
+        goto Done;
+    }
+
+    /* Get username pointer from options.
+     */
+
+    if (MI_DestinationOptions_GetCredentialsAt(pDestOptions, 0, (const MI_Char **)&optionName, &userCredentials, NULL) 
+          != MI_RESULT_OK)
+    {
+        // Log here
+        LOGE2((ZT("_UnpackDestinationOptions: No credentials available.")));
+        result = MI_RESULT_INVALID_PARAMETER;
+        goto Done;
+    }         
+
+
+    /* First delivery. pAuthType. We convert the string into an enum */
+    
+    method = AUTH_METHOD_NONE;
+    if (userCredentials.authenticationType)
+    {
+        method = AUTH_METHOD_UNSUPPORTED;
+    }
+
+    if (Tcscasecmp(userCredentials.authenticationType, AUTH_NAME_BASIC) == 0)
+    {
+        method = AUTH_METHOD_BASIC;
+    }
+   
+    if (Tcscasecmp(userCredentials.authenticationType, AUTH_NAME_NEGOTIATE) == 0) 
```

Comment:
should be else's here otherwise it is string comparing many times even if it already achieved a match. Should probably do a fast string match instead at some point.